pennylane_lightning/core/_adjoint_jacobian_base.py

```diff
@@ -0,0 +1,290 @@
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Internal methods for adjoint Jacobian differentiation method.
+"""
+
+from abc import ABC, abstractmethod
+from typing import Any, Callable, List
+
+import numpy as np
+import pennylane as qml
+from pennylane import BasisState, QuantumFunctionError, StatePrep
+from pennylane.measurements import Expectation, MeasurementProcess, State
+from pennylane.operation import Operation
+from pennylane.tape import QuantumTape
+
+from pennylane_lightning.core._serialize import QuantumScriptSerializer
+
+
+class LightningBaseAdjointJacobian(ABC):
+    """Lightning [Device] Adjoint Jacobian class
+
+    A class that serves as a base class for Lightning state-vector simulators.
+    Check and execute the adjoint Jacobian differentiation method.
+
+    Args:
+        qubit_state(Lightning[Device]StateVector): State Vector to calculate the adjoint Jacobian with.
+        batch_obs(bool): If serialized tape is to be batched or not.
+    """
+
+    def __init__(self, qubit_state: Any, batch_obs: bool) -> None:
+        self._qubit_state = qubit_state
+        self._batch_obs = batch_obs
+
+        # Dummy for the C++ bindings
+        self._jacobian_lightning: Callable = None
+        self._create_ops_list_lightning: Callable = None
+
+    @property
+    def qubit_state(self):
+        """Returns a handle to the Lightning[Device]StateVector object."""
+        return self._qubit_state
+
+    @property
+    def state(self):
+        """Returns a handle to the Lightning internal data object."""
+        return self._qubit_state.state_vector
+
+    @property
+    def dtype(self):
+        """Returns the simulation data type."""
+        return self._qubit_state.dtype
+
+    @abstractmethod
+    def _adjoint_jacobian_dtype(self):
+        """Binding to Lightning [Device] Adjoint Jacobian C++ class.
+
+        Returns: the AdjointJacobian class
+        """
+
+    @staticmethod
+    def _get_return_type(
+        measurements: List[MeasurementProcess],
+    ):
+        """Get the measurement return type.
+
+        Args:
+            measurements (List[MeasurementProcess]): a list of measurement processes to check.
+
+        Returns:
+            None, Expectation or State: a common return type of measurements.
+        """
+        if not measurements:
+            return None
+
+        if len(measurements) == 1 and measurements[0].return_type is State:
+            return State
+
+        return Expectation
+
+    def _process_jacobian_tape(
+        self, tape: QuantumTape, use_mpi: bool = False, split_obs: bool = False
```

Comment:
Not all devices have support for mpi.. this should be optional.
```suggestion
        self, tape: QuantumTape, split_obs: bool = False
```